### ISSUE TITLE
Crashed on UnknownHostException in Rx chain

### ISSUE THREAD

**Issue by consp1racy** [instacart/truetime-android#42]
TrueTime 2.2.0
Device: Sony F5321
Android: 7.0

```
Caused by java.net.UnknownHostException: Unable to resolve host "time.google.com": No address associated with hostname
       at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:125)
       at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:74)
       at java.net.InetAddress.getAllByName(InetAddress.java:752)
       at com.instacart.library.truetime.TrueTimeRx$4$2.call(TrueTimeRx.java:113)
       at com.instacart.library.truetime.TrueTimeRx$4$2.call(TrueTimeRx.java:108)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69)
       at rx.internal.util.ScalarSynchronousObservable$ScalarAsyncProducer.call(ScalarSynchronousObservable.java:200)
       at rx.internal.util.ScalarSynchronousObservable$2$1.call(ScalarSynchronousObservable.java:114)
       at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230)
       at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
       at java.lang.Thread.run(Thread.java:761)
Caused by android.system.GaiException: android_getaddrinfo failed: EAI_NODATA (No address associated with hostname)
       at libcore.io.Posix.android_getaddrinfo(Posix.java)
       at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:55)
       at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:106)
       at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:74)
       at java.net.InetAddress.getAllByName(InetAddress.java:752)
       at com.instacart.library.truetime.TrueTimeRx$4$2.call(TrueTimeRx.java:113)
       at com.instacart.library.truetime.TrueTimeRx$4$2.call(TrueTimeRx.java:108)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69)
       at rx.internal.util.ScalarSynchronousObservable$ScalarAsyncProducer.call(ScalarSynchronousObservable.java:200)
       at rx.internal.util.ScalarSynchronousObservable$2$1.call(ScalarSynchronousObservable.java:114)
       at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230)
       at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
       at java.lang.Thread.run(Thread.java:761)
```

**Comment by consp1racy** [comment]
Full stack trace:
```
Fatal Exception: java.lang.IllegalStateException: Fatal Exception thrown on Scheduler.Worker thread.
       at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:59)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
       at java.lang.Thread.run(Thread.java:761)
Caused by rx.exceptions.OnErrorFailedException: Error occurred when trying to propagate error to Observer.onError
       at rx.observers.SafeSubscriber._onError(SafeSubscriber.java:187)
       at rx.observers.SafeSubscriber.onError(SafeSubscriber.java:115)
       at rx.internal.operators.OperatorSubscribeOn$1$1.onError(OperatorSubscribeOn.java:59)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onError(OnSubscribeMap.java:88)
       at rx.internal.operators.OnSubscribeDoOnEach$DoOnEachSubscriber.onError(OnSubscribeDoOnEach.java:87)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onError(OnSubscribeMap.java:88)
       at rx.internal.operators.OnSubscribeFilter$FilterSubscriber.onError(OnSubscribeFilter.java:90)
       at rx.internal.operators.OperatorToObservableList$1.onError(OperatorToObservableList.java:104)
       at rx.internal.operators.OperatorTake$1.onError(OperatorTake.java:65)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.reportError(OperatorMerge.java:266)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.checkTerminate(OperatorMerge.java:818)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:579)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:568)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.onError(OperatorMerge.java:276)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onError(OnSubscribeMap.java:88)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onError(OnSubscribeMap.java:88)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.reportError(OperatorMerge.java:266)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.checkTerminate(OperatorMerge.java:818)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:579)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:568)
       at rx.internal.operators.OperatorMerge$InnerSubscriber.onError(OperatorMerge.java:852)
       at rx.internal.operators.OnSubscribeThrow.call(OnSubscribeThrow.java:44)
       at rx.internal.operators.OnSubscribeThrow.call(OnSubscribeThrow.java:28)
       at rx.Observable.unsafeSubscribe(Observable.java:10144)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:248)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:148)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:77)
       at rx.internal.util.ScalarSynchronousObservable$ScalarAsyncProducer.call(ScalarSynchronousObservable.java:200)
       at rx.internal.util.ScalarSynchronousObservable$2$1.call(ScalarSynchronousObservable.java:114)
       at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230)
       at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
       at java.lang.Thread.run(Thread.java:761)
Caused by rx.exceptions.CompositeException: 1 exceptions occurred. 
       at rx.observers.SafeSubscriber._onError(SafeSubscriber.java:187)
       at rx.observers.SafeSubscriber.onError(SafeSubscriber.java:115)
       at rx.internal.operators.OperatorSubscribeOn$1$1.onError(OperatorSubscribeOn.java:59)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onError(OnSubscribeMap.java:88)
       at rx.internal.operators.OnSubscribeDoOnEach$DoOnEachSubscriber.onError(OnSubscribeDoOnEach.java:87)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onError(OnSubscribeMap.java:88)
       at rx.internal.operators.OnSubscribeFilter$FilterSubscriber.onError(OnSubscribeFilter.java:90)
       at rx.internal.operators.OperatorToObservableList$1.onError(OperatorToObservableList.java:104)
       at rx.internal.operators.OperatorTake$1.onError(OperatorTake.java:65)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.reportError(OperatorMerge.java:266)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.checkTerminate(OperatorMerge.java:818)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:579)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:568)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.onError(OperatorMerge.java:276)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onError(OnSubscribeMap.java:88)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onError(OnSubscribeMap.java:88)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.reportError(OperatorMerge.java:266)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.checkTerminate(OperatorMerge.java:818)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:579)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:568)
       at rx.internal.operators.OperatorMerge$InnerSubscriber.onError(OperatorMerge.java:852)
       at rx.internal.operators.OnSubscribeThrow.call(OnSubscribeThrow.java:44)
       at rx.internal.operators.OnSubscribeThrow.call(OnSubscribeThrow.java:28)
       at rx.Observable.unsafeSubscribe(Observable.java:10144)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:248)
       at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:148)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:77)
       at rx.internal.util.ScalarSynchronousObservable$ScalarAsyncProducer.call(ScalarSynchronousObservable.java:200)
       at rx.internal.util.ScalarSynchronousObservable$2$1.call(ScalarSynchronousObservable.java:114)
       at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230)
       at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
       at java.lang.Thread.run(Thread.java:761)
Caused by rx.exceptions.CompositeException$CompositeExceptionCausalChain: Chain of Causes for CompositeException In Order Received =>
       at com.crashlytics.android.core.TrimmedThrowableData.(TrimmedThrowableData.java)
       at com.crashlytics.android.core.TrimmedThrowableData.(TrimmedThrowableData.java)
       at com.crashlytics.android.core.TrimmedThrowableData.(TrimmedThrowableData.java)
       at com.crashlytics.android.core.CrashlyticsController.writeSessionEvent(CrashlyticsController.java:1068)
       at com.crashlytics.android.core.CrashlyticsController.writeFatal(CrashlyticsController.java:831)
       at com.crashlytics.android.core.CrashlyticsController.access$200(CrashlyticsController.java:59)
       at com.crashlytics.android.core.CrashlyticsController$6.call(CrashlyticsController.java:285)
       at com.crashlytics.android.core.CrashlyticsController$6.call(CrashlyticsController.java:278)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
       at io.fabric.sdk.android.services.common.ExecutorUtils$1$1.onRun(ExecutorUtils.java:75)
       at io.fabric.sdk.android.services.common.BackgroundPriorityRunnable.run(BackgroundPriorityRunnable.java:30)
       at java.lang.Thread.run(Thread.java:761)
Caused by java.net.UnknownHostException: Unable to resolve host "time.google.com": No address associated with hostname
       at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:125)
       at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:74)
       at java.net.InetAddress.getAllByName(InetAddress.java:752)
       at com.instacart.library.truetime.TrueTimeRx$4$2.call(TrueTimeRx.java:113)
       at com.instacart.library.truetime.TrueTimeRx$4$2.call(TrueTimeRx.java:108)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69)
       at rx.internal.util.ScalarSynchronousObservable$ScalarAsyncProducer.call(ScalarSynchronousObservable.java:200)
       at rx.internal.util.ScalarSynchronousObservable$2$1.call(ScalarSynchronousObservable.java:114)
       at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230)
       at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
       at java.lang.Thread.run(Thread.java:761)
Caused by android.system.GaiException: android_getaddrinfo failed: EAI_NODATA (No address associated with hostname)
       at libcore.io.Posix.android_getaddrinfo(Posix.java)
       at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:55)
       at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:106)
       at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:74)
       at java.net.InetAddress.getAllByName(InetAddress.java:752)
       at com.instacart.library.truetime.TrueTimeRx$4$2.call(TrueTimeRx.java:113)
       at com.instacart.library.truetime.TrueTimeRx$4$2.call(TrueTimeRx.java:108)
       at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69)
       at rx.internal.util.ScalarSynchronousObservable$ScalarAsyncProducer.call(ScalarSynchronousObservable.java:200)
       at rx.internal.util.ScalarSynchronousObservable$2$1.call(ScalarSynchronousObservable.java:114)
       at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230)
       at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
       at java.lang.Thread.run(Thread.java:761)
```

**Comment by kaushikgopal** [comment]
we've been observing a bunch of these in production too. Internally i've been tracking a whole bunch of things we've tried to tackle this one. briefly:

* happens with a subset of android devices only. i could not find a pattern when looking at our internal data. it's very random: i do however notice, that it happens much more frequently when apps are on mobile data
* the crash is surfaced `onError`, so i imagine you're handling the surfaced exception (just so anyone else reading this issue understands :) )
* we've observed this with multiple ntp hosts, i.e. `time.google.com`, `time.apple.com`, so it isn't anything specific to one of the ntp hosts. thus switching ntp hosts won't solve this problem.
* cursory searching on stack overflow suggested turning wifi/airplane mode on/off to kickstart the network radio. we have had no luck with this
* what is even more odd about this, is that other network connections work, i.e. the user probably has network connectivity so it isn't a case of misconfigured permissions or network truly not being available.

My most recent attempt for solving this issue has been providing an ip address directly vs the ntp host domain name. i think we're having some luck with that but waiting on more data.

even if that solves the problem, it would obviously not be a long term solution but would atleast hint at what the problem could potentially be. will keep this issue posted as i gather more data.

if anyone reading this can consistently reproduce the issue, then please chime in. it would be super fast and easy to iterate to a solution faster, if we had a reproducible case.

**Comment by consp1racy** [comment]
I recall working around this by manually calling InetAddress.getAllBy once more before actually needing it and discarding the exception if any.

**Comment by kaushikgopal** [comment]
interesting i did notice some folks mentioning `InetAddress.getByName` causing some trouble but `getAllBy` worked fine.

so you're saying calling `InetAddress.getAllBy` a second time seems to work 😕  ? i would imagine the exception being called the second time as well right? 

the use of a direct IP (so far atleast) seems to be performing significantly better. 🤞 . too early to tell though.

**Comment by consp1racy** [comment]
Sorry, typo, `getbyName`.

**Comment by kaushikgopal** [comment]
hmm. stack overfllow says getAllByName is more reliable, but only way to confirm this is real testing. i've added a flag in that test branch. will try it and report back on results.

using the ip address seems to work (but is obv. not terribly feasible). so will see how the results compare here

**Comment by kaushikgopal** [comment]
after much wrestling with this issue, and numerous experiments, my initial suspicion seemed right. The dns resolution was to blame.

Supplying an ip address directly works well.

Further reading led me to this article which talks of a similar problem: https://garage.easytaxi.com/2015/12/11/tackling-dns-issues-on-android/

I've basically adopted a similar strategy to address the problem. TrueTimeRx now has an additional api that takes in a bunch of resolved IPs (see #338e1a8). So at Instacart, we use a custom DNS resolver (that's easily configurable on okHttp) and supply the resolved IPs directly to TrueTime.

I consciously chose NOT to do the Dns resolution inside TrueTime because 

a. i don't want additional dependencies on other libraries and 
b. at it's core, TrueTime shouldn't have to deal with DNS resolution. It merely tries to query for time from the IP address supplied.

I can post a snippet of the DNS resolver here if anyone wants, but it's very similar to the one mentioned in the pot by the easytaxi folks.
